File: core/src/main/java/com/uiptv/util/ImageUrlNormalizer.java
package com.uiptv.util;

import com.uiptv.model.Account;

import java.net.URI;
import java.util.List;

import static com.uiptv.util.StringUtils.isBlank;

public final class ImageUrlNormalizer {
    private ImageUrlNormalizer() {
    }

    public static String normalizeImageUrl(String imageUrl, Account account) {
        if (isBlank(imageUrl)) {
            return "";
        }
        String value = imageUrl.trim().replace("\\/", "/");
        value = trimWrappedImageQuotes(value);
        if (isBlank(value)) {
            return "";
        }
        if (isAbsoluteImageUrl(value) || isInlineImageUrl(value)) {
            return value;
        }
        URI base = resolveBaseUri(account);
        String scheme = resolveBaseScheme(base);
        String host = resolveBaseHost(base);
        int port = base == null ? -1 : base.getPort();
        if (value.startsWith("//")) {
            return scheme + ":" + value;
        }
        if (value.startsWith("/")) {
            return buildRootRelativeImageUrl(value, scheme, host, port);
        }
        if (value.matches("^[a-zA-Z0-9.-]+(?::\\d+)?/.*") && isBlank(host)) {
            int slash = value.indexOf('/');
            String hostCandidate = slash > 0 ? value.substring(0, slash) : value;
            if (hostCandidate.contains(".") || "localhost".equalsIgnoreCase(hostCandidate)) {
                return scheme + "://" + value;
            }
        }
        return buildRelativeImageUrl(value, scheme, host, port);
    }

    private static URI resolveBaseUri(Account account) {
        if (account == null) {
            return null;
        }
        List<String> candidates = new java.util.ArrayList<>();
        if (!isBlank(account.getServerPortalUrl())) {
            candidates.add(account.getServerPortalUrl());
        }
        if (!isBlank(account.getUrl())) {
            candidates.add(account.getUrl());
        }
        for (String candidate : candidates) {
            URI resolved = parseCandidateBaseUri(candidate);
            if (resolved != null) {
                return resolved;
            }
        }
        return null;
    }

    private static String resolveBaseScheme(URI base) {
        return base != null && !isBlank(base.getScheme()) ? base.getScheme() : "https";
    }

    private static String resolveBaseHost(URI base) {
        return base != null && !isBlank(base.getHost()) ? base.getHost() : "";
    }

    private static String trimWrappedImageQuotes(String value) {
        if (isBlank(value)) {
            return "";
        }
        String trimmed = value.trim();
        if (trimmed.length() < 2) {
            return trimmed;
        }
        if ((trimmed.startsWith("\"") && trimmed.endsWith("\""))
                || (trimmed.startsWith("'") && trimmed.endsWith("'"))) {
            return trimmed.substring(1, trimmed.length() - 1).trim();
        }
        return trimmed;
    }

    private static boolean isAbsoluteImageUrl(String value) {
        return value.matches("^[a-zA-Z][a-zA-Z0-9+.-]*://.*");
    }

    private static boolean isInlineImageUrl(String value) {
        return value.startsWith("data:") || value.startsWith("blob:") || value.startsWith("file:");
    }

    private static String buildRootRelativeImageUrl(String value, String scheme, String host, int port) {
        if (!isBlank(host)) {
            return scheme + "://" + host + formatPort(port) + value;
        }
        return value;
    }

    private static String buildRelativeImageUrl(String value, String scheme, String host, int port) {
        String normalized = value.startsWith("./") ? value.substring(2) : value;
        if (!isBlank(host)) {
            return scheme + "://" + host + formatPort(port) + "/" + normalized;
        }
        return ServerUrlUtil.getLocalServerUrl() + "/" + normalized;
    }

    private static String formatPort(int port) {
        return port > 0 ? ":" + port : "";
    }

    private static URI parseCandidateBaseUri(String candidate) {
        if (isBlank(candidate)) {
            return null;
        }
        try {
            URI uri = URI.create(candidate.trim());
            if (!isBlank(uri.getHost())) {
                return uri;
            }
            if (isBlank(uri.getScheme())) {
                URI withScheme = URI.create("http://" + candidate.trim());
                if (!isBlank(withScheme.getHost())) {
                    return withScheme;
                }
            }
        } catch (Exception _) {
            // Invalid image/base URIs should not break rendering.
        }
        return null;
    }
}

File: server/src/main/kotlin/com/uiptv/util/ImageUrlNormalizer.kt
package com.uiptv.util

import com.uiptv.model.Account
import java.net.URI

object ImageUrlNormalizer {
    @JvmStatic
    fun normalizeImageUrl(imageUrl: String?, account: Account?): String {
        if (StringUtils.isBlank(imageUrl)) {
            return ""
        }
        var value = imageUrl!!.trim().replace("\\/", "/")
        value = trimWrappedImageQuotes(value)
        if (StringUtils.isBlank(value)) {
            return ""
        }
        if (isAbsoluteImageUrl(value) || isInlineImageUrl(value)) {
            return value
        }
        val base = resolveBaseUri(account)
        val scheme = resolveBaseScheme(base)
        val host = resolveBaseHost(base)
        val port = base?.port ?: -1
        if (value.startsWith("//")) {
            return "$scheme:$value"
        }
        if (value.startsWith("/")) {
            return buildRootRelativeImageUrl(value, scheme, host, port)
        }
        if (value.matches(Regex("^[a-zA-Z0-9.-]+(?::\\d+)?/.*")) && StringUtils.isBlank(host)) {
            val slash = value.indexOf('/')
            val hostCandidate = if (slash > 0) value.substring(0, slash) else value
            if (hostCandidate.contains(".") || hostCandidate.equals("localhost", ignoreCase = true)) {
                return "$scheme://$value"
            }
        }
        return buildRelativeImageUrl(value, scheme, host, port)
    }

    private fun resolveBaseUri(account: Account?): URI? {
        if (account == null) {
            return null
        }
        val candidates = ArrayList<String>()
        if (!StringUtils.isBlank(account.serverPortalUrl)) {
            candidates.add(account.serverPortalUrl!!)
        }
        if (!StringUtils.isBlank(account.url)) {
            candidates.add(account.url!!)
        }
        for (candidate in candidates) {
            val resolved = parseCandidateBaseUri(candidate)
            if (resolved != null) {
                return resolved
            }
        }
        return null
    }

    private fun resolveBaseScheme(base: URI?): String =
        if (base != null && !StringUtils.isBlank(base.scheme)) base.scheme else "https"

    private fun resolveBaseHost(base: URI?): String =
        if (base != null && !StringUtils.isBlank(base.host)) base.host else ""

    private fun trimWrappedImageQuotes(value: String): String {
        if (StringUtils.isBlank(value)) {
            return ""
        }
        val trimmed = value.trim()
        if (trimmed.length < 2) {
            return trimmed
        }
        return if ((trimmed.startsWith("\"") && trimmed.endsWith("\"")) || (trimmed.startsWith("'") && trimmed.endsWith("'"))) {
            trimmed.substring(1, trimmed.length - 1).trim()
        } else {
            trimmed
        }
    }

    private fun isAbsoluteImageUrl(value: String): Boolean =
        value.matches(Regex("^[a-zA-Z][a-zA-Z0-9+.-]*://.*"))

    private fun isInlineImageUrl(value: String): Boolean =
        value.startsWith("data:") || value.startsWith("blob:") || value.startsWith("file:")

    private fun buildRootRelativeImageUrl(value: String, scheme: String, host: String, port: Int): String {
        return if (!StringUtils.isBlank(host)) {
            "$scheme://$host${formatPort(port)}$value"
        } else {
            value
        }
    }

    private fun buildRelativeImageUrl(value: String, scheme: String, host: String, port: Int): String {
        val normalized = if (value.startsWith("./")) value.substring(2) else value
        return if (!StringUtils.isBlank(host)) {
            "$scheme://$host${formatPort(port)}/$normalized"
        } else {
            ServerUrlUtil.getLocalServerUrl() + "/$normalized"
        }
    }

    private fun formatPort(port: Int): String = if (port > 0) ":$port" else ""

    private fun parseCandidateBaseUri(candidate: String?): URI? {
        if (StringUtils.isBlank(candidate)) {
            return null
        }
        return try {
            val uri = URI.create(candidate!!.trim())
            if (!StringUtils.isBlank(uri.host)) {
                return uri
            }
            if (StringUtils.isBlank(uri.scheme)) {
                val withScheme = URI.create("http://${candidate.trim()}")
                if (!StringUtils.isBlank(withScheme.host)) {
                    return withScheme
                }
            }
            null
        } catch (_: Exception) {
            null
        }
    }
}
